File: src/dotnet/TraceMap.Core/SqlShapeExtractor.cs
using System.Text.RegularExpressions;

namespace TraceMap.Core;

public sealed record SqlQueryShape(
    string OperationName,
    IReadOnlyList<string> TableNames,
    IReadOnlyList<string> ColumnNames,
    string QueryShapeHash)
{
    public string? PrimaryTable => TableNames.Count == 0 ? null : TableNames[0];
}

public static class SqlShapeExtractor
{
    private static readonly HashSet<string> SqlVerbs = new(StringComparer.Ordinal)
    {
        "SELECT",
        "INSERT",
        "UPDATE",
        "DELETE",
        "MERGE",
        "CREATE",
        "ALTER",
        "DROP",
        "TRUNCATE",
        "CALL",
        "EXEC",
        "EXECUTE"
    };

    private static readonly HashSet<string> SqlStopWords = new(StringComparer.Ordinal)
    {
        "AND",
        "AS",
        "ASC",
        "BETWEEN",
        "BY",
        "CASE",
        "DESC",
        "DISTINCT",
        "ELSE",
        "END",
        "FALSE",
        "FROM",
        "GROUP",
        "HAVING",
        "IN",
        "IS",
        "JOIN",
        "LEFT",
        "LIKE",
        "LIMIT",
        "NOT",
        "NULL",
        "ON",
        "OR",
        "ORDER",
        "RIGHT",
        "SELECT",
        "SET",
        "THEN",
        "TRUE",
        "VALUES",
        "WHEN",
        "WHERE"
    };

    public static bool IsSqlLike(string value)
    {
        var first = FirstToken(value);
        return SqlVerbs.Contains(first) || first == "WITH";
    }

    public static string OperationName(string value)
    {
        var first = FirstToken(value);
        return SqlVerbs.Contains(first) ? first : string.Empty;
    }

    public static SqlQueryShape QueryShape(string value)
    {
        var normalized = NormalizeSql(value);
        var operation = ShapeOperation(normalized);
        var tables = TableNames(normalized, operation);
        var columns = ColumnNames(normalized, operation);
        return new SqlQueryShape(operation, tables, columns, FactFactory.Hash(normalized, 32));
    }

    public static SortedDictionary<string, string> QueryShapeProperties(string value, string sourceKind)
    {
        var shape = QueryShape(value);
        var properties = new SortedDictionary<string, string>(StringComparer.Ordinal)
        {
            ["textHash"] = FactFactory.Hash(value, 32),
            ["queryShapeHash"] = shape.QueryShapeHash,
            ["sqlSourceKind"] = sourceKind
        };
        if (!string.IsNullOrWhiteSpace(shape.OperationName))
        {
            properties["operationName"] = shape.OperationName;
        }

        if (shape.PrimaryTable is { Length: > 0 } primary)
        {
            properties["tableName"] = primary;
        }

        if (shape.TableNames.Count > 0)
        {
            properties["tableNames"] = string.Join(';', shape.TableNames);
        }

        if (shape.ColumnNames.Count > 0)
        {
            var columns = string.Join(';', shape.ColumnNames);
            properties["columnNames"] = columns;
            properties["fieldNames"] = columns;
        }

        return properties;
    }

    internal static string NormalizeSql(string value)
    {
        value = Regex.Replace(value, "--[^\n\r]*", " ", RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100));
        value = Regex.Replace(value, @"/\*.*?\*/", " ", RegexOptions.CultureInvariant | RegexOptions.Singleline, TimeSpan.FromMilliseconds(100));
        value = Regex.Replace(value, @"'(?:''|\\['""]|[^'])*'", "' '", RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100));
        value = Regex.Replace(value, @"""(?:""""|\\[""']|[^""])*""", "\" \"", RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100));
        value = Regex.Replace(value, @"\s+", " ", RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100));
        return value.Trim().TrimEnd(';');
    }

    private static string ShapeOperation(string value)
    {
        var first = FirstToken(value);
        return SqlVerbs.Contains(first) ? first : string.Empty;
    }

    private static string FirstToken(string value)
    {
        var trimmed = value.TrimStart();
        if (trimmed.Length == 0)
        {
            return string.Empty;
        }

        var match = Regex.Match(trimmed, @"\S+", RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100));
        return match.Success ? match.Value.ToUpperInvariant() : string.Empty;
    }

    private static IReadOnlyList<string> TableNames(string sql, string operation)
    {
        var candidates = new List<string>();
        switch (operation)
        {
            case "SELECT":
                candidates.AddRange(TopLevelMatches(sql, @"\bFROM\s+([A-Za-z_][A-Za-z0-9_.$\[\]""`]*)"));
                candidates.AddRange(TopLevelMatches(sql, @"\bJOIN\s+([A-Za-z_][A-Za-z0-9_.$\[\]""`]*)"));
                break;
            case "INSERT":
                candidates.AddRange(Matches(sql, @"\bINSERT\s+INTO\s+([A-Za-z_][A-Za-z0-9_.$\[\]""`]*)"));
                break;
            case "UPDATE":
                candidates.AddRange(Matches(sql, @"\bUPDATE\s+([A-Za-z_][A-Za-z0-9_.$\[\]""`]*)"));
                break;
            case "DELETE":
                candidates.AddRange(Matches(sql, @"\bDELETE\s+FROM\s+([A-Za-z_][A-Za-z0-9_.$\[\]""`]*)"));
                break;
            case "CREATE":
                candidates.AddRange(Matches(sql, @"\bCREATE\s+(?:TEMP(?:ORARY)?\s+)?TABLE\s+(?:IF\s+NOT\s+EXISTS\s+)?([A-Za-z_][A-Za-z0-9_.$\[\]""`]*)"));
                break;
            case "DROP":
            case "TRUNCATE":
            case "ALTER":
                candidates.AddRange(Matches(sql, $@"\b{operation}\s+(?:TABLE\s+)?([A-Za-z_][A-Za-z0-9_.$\[\]""`]*)"));
                break;
        }

        // CALL/EXEC routine names are intentionally not table candidates in v1.
        return Unique(candidates.Select(CleanIdentifier));
    }

    private static IReadOnlyList<string> ColumnNames(string sql, string operation)
    {
        return operation switch
        {
            "SELECT" => SelectColumns(Between(sql, "SELECT", "FROM")),
            "INSERT" => SplitIdentifierList(MatchGroup(sql, @"\bINSERT\s+INTO\s+[A-Za-z_][A-Za-z0-9_.$\[\]""`]*\s*\(([^)]*)\)")),
            "UPDATE" => Unique(SplitCsv(Between(sql, "SET", "WHERE")).Select(part => CleanIdentifier(part.Split('=', 2)[0]))),
            "CREATE" => CreateTableColumns(MatchGroup(sql, @"\bCREATE\s+(?:TEMP(?:ORARY)?\s+)?TABLE\s+(?:IF\s+NOT\s+EXISTS\s+)?[A-Za-z_][A-Za-z0-9_.$\[\]""`]*\s*\((.*)\)", RegexOptions.IgnoreCase | RegexOptions.Singleline)),
            _ => []
        };
    }

    private static IReadOnlyList<string> Matches(string sql, string pattern)
    {
        return Regex.Matches(sql, pattern, RegexOptions.IgnoreCase | RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100))
            .Select(match => match.Groups[1].Value)
            .ToArray();
    }

    private static IReadOnlyList<string> TopLevelMatches(string sql, string pattern)
    {
        return Regex.Matches(sql, pattern, RegexOptions.IgnoreCase | RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100))
            .Where(match => ParenthesisDepthBefore(sql, match.Index) == 0)
            .Select(match => match.Groups[1].Value)
            .ToArray();
    }

    private static int ParenthesisDepthBefore(string sql, int index)
    {
        var depth = 0;
        for (var offset = 0; offset < index && offset < sql.Length; offset++)
        {
            if (sql[offset] == '(')
            {
                depth++;
            }
            else if (sql[offset] == ')' && depth > 0)
            {
                depth--;
            }
        }

        return depth;
    }

    private static string MatchGroup(string sql, string pattern, RegexOptions options = RegexOptions.IgnoreCase)
    {
        var match = Regex.Match(sql, pattern, options | RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100));
        return match.Success ? match.Groups[1].Value : string.Empty;
    }

    private static string Between(string sql, string start, string end)
    {
        var match = Regex.Match(sql, $@"\b{start}\b(.*?)(?:\b{end}\b|$)", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant | RegexOptions.Singleline, TimeSpan.FromMilliseconds(100));
        return match.Success ? match.Groups[1].Value : string.Empty;
    }

    private static IReadOnlyList<string> SelectColumns(string text)
    {
        var columns = new List<string>();
        foreach (var part in SplitCsv(text))
        {
            var cleaned = Regex.Replace(part.Trim(), @"\bAS\b\s+[A-Za-z_][A-Za-z0-9_]*$", string.Empty, RegexOptions.IgnoreCase | RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100));
            var token = cleaned.Contains('.') ? cleaned[(cleaned.LastIndexOf('.') + 1)..].Trim() : cleaned.Trim();
            if (Regex.IsMatch(token, @"\s", RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100)))
            {
                token = Regex.Split(token, @"\s+").Last();
            }

            var name = CleanIdentifier(token);
            if (!string.IsNullOrWhiteSpace(name)
                && name != "*"
                && !SqlStopWords.Contains(name.ToUpperInvariant())
                && Regex.IsMatch(name, @"^[A-Za-z_][A-Za-z0-9_]*$", RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100)))
            {
                columns.Add(name);
            }
        }

        return Unique(columns);
    }

    private static IReadOnlyList<string> CreateTableColumns(string text)
    {
        var columns = new List<string>();
        foreach (var part in SplitCsv(text))
        {
            var trimmed = part.Trim();
            var first = trimmed.Length == 0 ? string.Empty : trimmed.Split((char[]?)null, 2, StringSplitOptions.RemoveEmptyEntries).FirstOrDefault() ?? string.Empty;
            var name = CleanIdentifier(first);
            if (!string.IsNullOrWhiteSpace(name)
                && !new[] { "CONSTRAINT", "PRIMARY", "FOREIGN", "UNIQUE", "CHECK", "KEY" }.Contains(name.ToUpperInvariant(), StringComparer.Ordinal))
            {
                columns.Add(name);
            }
        }

        return Unique(columns);
    }

    private static IReadOnlyList<string> SplitIdentifierList(string text)
    {
        return Unique(SplitCsv(text).Select(CleanIdentifier));
    }

    private static IReadOnlyList<string> SplitCsv(string text)
    {
        var parts = new List<string>();
        var depth = 0;
        var start = 0;
        for (var index = 0; index < text.Length; index++)
        {
            var current = text[index];
            if (current == '(')
            {
                depth++;
            }
            else if (current == ')' && depth > 0)
            {
                depth--;
            }
            else if (current == ',' && depth == 0)
            {
                parts.Add(text[start..index].Trim());
                start = index + 1;
            }
        }

        var tail = text[start..].Trim();
        if (tail.Length > 0)
        {
            parts.Add(tail);
        }

        return parts;
    }

    private static string CleanIdentifier(string value)
    {
        value = value.Trim().Trim(',', ';').Trim('"', '`', '[', ']');
        if (value.Contains('.', StringComparison.Ordinal))
        {
            value = value[(value.LastIndexOf('.') + 1)..].Trim('"', '`', '[', ']');
        }

        if (value.Length == 0 || !Regex.IsMatch(value, @"^[A-Za-z_][A-Za-z0-9_]*$", RegexOptions.CultureInvariant, TimeSpan.FromMilliseconds(100)))
        {
            return string.Empty;
        }

        return value;
    }

    private static IReadOnlyList<string> Unique(IEnumerable<string> values)
    {
        var seen = new HashSet<string>(StringComparer.Ordinal);
        var result = new List<string>();
        foreach (var value in values)
        {
            if (!string.IsNullOrWhiteSpace(value) && seen.Add(value))
            {
                result.Add(value);
            }
        }

        return result;
    }
}

File: src/jvm/src/main/java/com/tracemap/jvm/extract/SqlShapeExtractor.java
package com.tracemap.jvm.extract;

import com.tracemap.jvm.util.Hashes;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Locale;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class SqlShapeExtractor {
    private static final Set<String> SQL_VERBS = Set.of("SELECT", "INSERT", "UPDATE", "DELETE", "MERGE", "CREATE", "ALTER", "DROP", "TRUNCATE", "CALL", "EXEC", "EXECUTE");
    private static final Set<String> SQL_STOP_WORDS = Set.of(
        "AND", "AS", "ASC", "BETWEEN", "BY", "CASE", "DESC", "DISTINCT", "ELSE", "END", "FALSE",
        "FROM", "GROUP", "HAVING", "IN", "IS", "JOIN", "LEFT", "LIKE", "LIMIT", "NOT", "NULL",
        "ON", "OR", "ORDER", "RIGHT", "SELECT", "SET", "THEN", "TRUE", "VALUES", "WHEN", "WHERE");

    private SqlShapeExtractor() {
    }

    public record Shape(String operationName, List<String> tableNames, List<String> columnNames, String queryShapeHash) {
        public String primaryTable() {
            return tableNames.isEmpty() ? "" : tableNames.get(0);
        }
    }

    public static boolean isSqlLike(String value) {
        String first = firstToken(value);
        return SQL_VERBS.contains(first) || "WITH".equals(first);
    }

    public static String operationName(String value) {
        String first = firstToken(value);
        return SQL_VERBS.contains(first) ? first : "";
    }

    public static Shape queryShape(String value) {
        String normalized = normalizeSql(value);
        String operation = shapeOperation(normalized);
        return new Shape(operation, tableNames(normalized, operation), columnNames(normalized, operation), Hashes.sha256(normalized, 32));
    }

    public static Map<String, String> queryShapeProperties(String value, String sourceKind) {
        Shape shape = queryShape(value);
        Map<String, String> props = new LinkedHashMap<>();
        props.put("textHash", Hashes.sha256(value, 32));
        props.put("queryShapeHash", shape.queryShapeHash());
        props.put("sqlSourceKind", sourceKind);
        if (!shape.operationName().isBlank()) {
            props.put("operationName", shape.operationName());
        }
        if (!shape.primaryTable().isBlank()) {
            props.put("tableName", shape.primaryTable());
        }
        if (!shape.tableNames().isEmpty()) {
            props.put("tableNames", String.join(";", shape.tableNames()));
        }
        if (!shape.columnNames().isEmpty()) {
            props.put("columnNames", String.join(";", shape.columnNames()));
            props.put("fieldNames", String.join(";", shape.columnNames()));
        }
        return props;
    }

    public static String normalizeSql(String value) {
        value = value.replaceAll("--[^\\n\\r]*", " ");
        value = Pattern.compile("/\\*.*?\\*/", Pattern.DOTALL).matcher(value).replaceAll(" ");
        value = value.replaceAll("'(?:''|\\\\['\"]|[^'])*'", "' '");
        value = value.replaceAll("\"(?:\"\"|\\\\[\"']|[^\"])*\"", "\" \"");
        value = value.replaceAll("\\s+", " ").trim();
        return value.replaceAll(";+\\z", "");
    }

    private static String shapeOperation(String value) {
        String first = firstToken(value);
        return SQL_VERBS.contains(first) ? first : "";
    }

    private static String firstToken(String value) {
        String trimmed = value.stripLeading();
        if (trimmed.isEmpty()) {
            return "";
        }
        return trimmed.split("\\s+", 2)[0].toUpperCase(Locale.ROOT);
    }

    private static List<String> tableNames(String sql, String operation) {
        List<String> candidates = new ArrayList<>();
        switch (operation) {
            case "SELECT" -> {
                candidates.addAll(matches(sql, "\\bFROM\\s+([A-Za-z_][A-Za-z0-9_.$\\[\\]\"`]*)"));
                candidates.addAll(matches(sql, "\\bJOIN\\s+([A-Za-z_][A-Za-z0-9_.$\\[\\]\"`]*)"));
            }
            case "INSERT" -> candidates.addAll(matches(sql, "\\bINSERT\\s+INTO\\s+([A-Za-z_][A-Za-z0-9_.$\\[\\]\"`]*)"));
            case "UPDATE" -> candidates.addAll(matches(sql, "\\bUPDATE\\s+([A-Za-z_][A-Za-z0-9_.$\\[\\]\"`]*)"));
            case "DELETE" -> candidates.addAll(matches(sql, "\\bDELETE\\s+FROM\\s+([A-Za-z_][A-Za-z0-9_.$\\[\\]\"`]*)"));
            case "CREATE" -> candidates.addAll(matches(sql, "\\bCREATE\\s+(?:TEMP(?:ORARY)?\\s+)?TABLE\\s+(?:IF\\s+NOT\\s+EXISTS\\s+)?([A-Za-z_][A-Za-z0-9_.$\\[\\]\"`]*)"));
            case "DROP", "TRUNCATE", "ALTER" -> candidates.addAll(matches(sql, "\\b" + operation + "\\s+(?:TABLE\\s+)?([A-Za-z_][A-Za-z0-9_.$\\[\\]\"`]*)"));
            default -> {
            }
        }
        return unique(candidates.stream().map(SqlShapeExtractor::cleanIdentifier).toList());
    }

    private static List<String> columnNames(String sql, String operation) {
        return switch (operation) {
            case "SELECT" -> selectColumns(between(sql, "SELECT", "FROM"));
            case "INSERT" -> splitIdentifierList(matchGroup(sql, "\\bINSERT\\s+INTO\\s+[A-Za-z_][A-Za-z0-9_.$\\[\\]\"`]*\\s*\\(([^)]*)\\)", 0));
            case "UPDATE" -> unique(splitCsv(between(sql, "SET", "WHERE")).stream().map(part -> cleanIdentifier(part.split("=", 2)[0])).toList());
            case "CREATE" -> createTableColumns(matchGroup(sql, "\\bCREATE\\s+(?:TEMP(?:ORARY)?\\s+)?TABLE\\s+(?:IF\\s+NOT\\s+EXISTS\\s+)?[A-Za-z_][A-Za-z0-9_.$\\[\\]\"`]*\\s*\\((.*)\\)", Pattern.DOTALL));
            default -> List.of();
        };
    }

    private static List<String> matches(String sql, String pattern) {
        Matcher matcher = Pattern.compile(pattern, Pattern.CASE_INSENSITIVE).matcher(sql);
        List<String> result = new ArrayList<>();
        while (matcher.find()) {
            result.add(matcher.group(1));
        }
        return result;
    }

    private static String matchGroup(String sql, String pattern, int flags) {
        Matcher matcher = Pattern.compile(pattern, Pattern.CASE_INSENSITIVE | flags).matcher(sql);
        return matcher.find() ? matcher.group(1) : "";
    }

    private static String between(String sql, String start, String end) {
        Matcher matcher = Pattern.compile("\\b" + start + "\\b(.*?)(?:\\b" + end + "\\b|$)", Pattern.CASE_INSENSITIVE | Pattern.DOTALL).matcher(sql);
        return matcher.find() ? matcher.group(1) : "";
    }

    private static List<String> selectColumns(String text) {
        List<String> columns = new ArrayList<>();
        for (String part : splitCsv(text)) {
            String cleaned = part.trim().replaceAll("(?i)\\bAS\\b\\s+[A-Za-z_][A-Za-z0-9_]*$", "");
            String token = cleaned.contains(".") ? cleaned.substring(cleaned.lastIndexOf('.') + 1).trim() : cleaned.trim();
            if (Pattern.compile("\\s").matcher(token).find()) {
                String[] pieces = token.split("\\s+");
                token = pieces.length == 0 ? "" : pieces[pieces.length - 1];
            }
            String name = cleanIdentifier(token);
            if (!name.isBlank() && !"*".equals(name) && !SQL_STOP_WORDS.contains(name.toUpperCase(Locale.ROOT)) && name.matches("^[A-Za-z_][A-Za-z0-9_]*$")) {
                columns.add(name);
            }
        }
        return unique(columns);
    }

    private static List<String> createTableColumns(String text) {
        List<String> columns = new ArrayList<>();
        for (String part : splitCsv(text)) {
            String[] pieces = part.trim().split("\\s+", 2);
            String name = cleanIdentifier(pieces.length == 0 ? "" : pieces[0]);
            if (!name.isBlank() && !Set.of("CONSTRAINT", "PRIMARY", "FOREIGN", "UNIQUE", "CHECK", "KEY").contains(name.toUpperCase(Locale.ROOT))) {
                columns.add(name);
            }
        }
        return unique(columns);
    }

    private static List<String> splitIdentifierList(String text) {
        return unique(splitCsv(text).stream().map(SqlShapeExtractor::cleanIdentifier).toList());
    }

    private static List<String> splitCsv(String text) {
        List<String> parts = new ArrayList<>();
        int depth = 0;
        int start = 0;
        for (int index = 0; index < text.length(); index++) {
            char current = text.charAt(index);
            if (current == '(') {
                depth++;
            } else if (current == ')' && depth > 0) {
                depth--;
            } else if (current == ',' && depth == 0) {
                parts.add(text.substring(start, index).trim());
                start = index + 1;
            }
        }
        String tail = text.substring(start).trim();
        if (!tail.isBlank()) {
            parts.add(tail);
        }
        return parts;
    }

    private static String cleanIdentifier(String value) {
        String cleaned = value.trim().replaceAll("^[,;]+|[,;]+$", "").replaceAll("^[\"`\\[\\]]+|[\"`\\[\\]]+$", "");
        if (cleaned.contains(".")) {
            cleaned = cleaned.substring(cleaned.lastIndexOf('.') + 1).replaceAll("^[\"`\\[\\]]+|[\"`\\[\\]]+$", "");
        }
        return cleaned.matches("^[A-Za-z_][A-Za-z0-9_]*$") ? cleaned : "";
    }

    private static List<String> unique(List<String> values) {
        LinkedHashSet<String> seen = new LinkedHashSet<>();
        for (String value : values) {
            if (value != null && !value.isBlank()) {
                seen.add(value);
            }
        }
        return new ArrayList<>(seen);
    }
}
